hbase-server/src/main/java/org/apache/hadoop/hbase/mob/MobFileCleanerChore.java

```diff
@@ -0,0 +1,310 @@
+        if (hcd.isMobEnabled() && hcd.getMinVersions() == 0) {
+          try {
+            cleaner.cleanExpiredMobFiles(htd.getTableName().getNameAsString(), hcd);
+          } catch (IOException e) {
+            LOG.error("Failed to clean the expired mob files table=" + htd.getTableName()
+                + " family=" + hcd.getNameAsString(), e);
+          }
+        }
+      }
+      try {
+        // Now clean obsolete files for a table
+        LOG.info("Cleaning obsolete MOB files from table={}", htd.getTableName());
+        cleanupObsoleteMobFiles(master.getConfiguration(), htd.getTableName());
+        LOG.info("Cleaning obsolete MOB files finished for table={}", htd.getTableName());
+      } catch (IOException e) {
+        LOG.error("Failed to clean the obsolete mob files for table=" + htd.getTableName(), e);
+      }
+    }
+  }
+
+  /**
+   * Performs housekeeping file cleaning (called by MOB Cleaner chore)
+   * @param conf configuration
+   * @param table table name
+   * @throws IOException exception
+   */
+  public void cleanupObsoleteMobFiles(Configuration conf, TableName table) throws IOException {
+
+    long minAgeToArchive =
+        conf.getLong(MobConstants.MIN_AGE_TO_ARCHIVE_KEY, MobConstants.DEFAULT_MIN_AGE_TO_ARCHIVE);
+    // We check only those MOB files, which creation time is less
+    // than maxCreationTimeToArchive. This is a current time - 1h. 1 hour gap
+    // gives us full confidence that all corresponding store files will
+    // exist at the time cleaning procedure begins and will be examined.
+    // So, if MOB file creation time is greater than this maxTimeToArchive,
+    // this will be skipped and won't be archived.
+    long maxCreationTimeToArchive = EnvironmentEdgeManager.currentTime() - minAgeToArchive;
+    LOG.info("Only MOB files whose creation time older than {} will be archived",
+      maxCreationTimeToArchive);
+    try (final Connection conn = ConnectionFactory.createConnection(conf);
+        final Admin admin = conn.getAdmin();) {
+      TableDescriptor htd = admin.getDescriptor(table);
+      List<ColumnFamilyDescriptor> list = MobUtils.getMobColumnFamilies(htd);
+      if (list.size() == 0) {
+        LOG.info("Skipping non-MOB table [{}]", table);
+        return;
+      }
+
+      Path rootDir = FSUtils.getRootDir(conf);
+      Path tableDir = FSUtils.getTableDir(rootDir, table);
+      // How safe is this call?
+      List<Path> regionDirs = FSUtils.getRegionDirs(FileSystem.get(conf), tableDir);
+
+      Set<String> allActiveMobFileName = new HashSet<String>();
+      FileSystem fs = FileSystem.get(conf);
+      for (Path regionPath : regionDirs) {
+        for (ColumnFamilyDescriptor hcd : list) {
+          String family = hcd.getNameAsString();
+          Path storePath = new Path(regionPath, family);
+          boolean succeed = false;
+          Set<String> regionMobs = new HashSet<String>();
+          while (!succeed) {
+            if (!fs.exists(storePath)) {
```

Comment:
New code:
```
      for (Path regionPath : regionDirs) {
        region:
        for (ColumnFamilyDescriptor hcd : list) {
          String family = hcd.getNameAsString();
          Path storePath = new Path(regionPath, family);
          boolean succeed = false;
          Set<String> regionMobs = new HashSet<String>();
          while (!succeed) {
            if (!fs.exists(storePath)) {
              LOG.warn("Directory {} was deleted during cleaner procedure execution," 
                  +" skipping region {}", storePath, regionPath.getName());
              continue region;
            }
```

If we find missing cf directory, we continue next region. MOB files are per region, so we can safely continue cleaning other regions.

Open question though. I will think about this.

Reply:
As far as I remember we disable Mob cleaner chore until migration is complete?

Reply:
On a potential issue with optimized compaction and old large MOB file ...
There is no issue here. The only difference between old and new MOB files is the new one has region name appended as a suffix. This is the artifact we no longer use. It was designed this way to identify MOB files belonging to a particular region in MOB cleaner chore. But now we rely on metadata in a store files to collect all active MOB files. So, basically we do not need these region suffixes anymore, but I will keep it this way. So, if old MOB is large that threshold during compaction run in optimized IO mode it is safe to keep it untouched. During migration (upgrade) processionals users will be able to select both: standard and optimized compaction.

Reply:
To make everything as smooth as possible, I will change the code in MobCompactor to detect old MOB files and disable optimized mode in this case.

Reply:
"For example, if I have 3 families "a", "b", and "c" and family "b" is deleted while in this loop then we'll continue to the next region and never account for the hfiles in family "c" that may have references to mob files."

I think we should abort cleaner in case of any IO failure - this is the most safest approach. We should be here on a conservative side - the optimization itself is not worth potential data loss. I will update the code.